storey/writers.py

```diff
@@ -15,58 +15,81 @@
-                    col, metadata_attr = col.split('=$', maxsplit=1)
-                    self._metadata_columns[col] = metadata_attr
-                elif '=' in col:
-                    col, rename_from = col.split('=', maxsplit=1)
-                    self._rename_columns[col] = rename_from
-                self._columns.append(col)
+            for column in columns:
+                if column in metadata_columns:
+                    metadata_attr = metadata_columns[column]
+                    new_value = getattr(event, metadata_attr)
+                elif column in rename_columns:
+                    new_value = event.body[rename_columns[column]]
+                else:
+                    new_value = event.body[column]
+                new_data.append(new_value)
+
+    @staticmethod
+    def _get_column_data_from_list(new_data, event, original_data, columns, metadata_columns):
+        data_cursor = 0
+        for column in columns:
+            if column in metadata_columns:
+                metadata_attr = metadata_columns[column]
+                new_value = getattr(event, metadata_attr)
+                new_data.append(new_value)
+            else:
+                new_data.append(original_data[data_cursor])
+                data_cursor += 1
+        return data_cursor
 
     def _event_to_writer_entry(self, event):
         data = event.body
         if isinstance(data, dict):
-            if self._first_event and self._infer_columns_from_data:
+            if self._infer_columns_from_data:
                 self._columns.extend(data.keys())
                 self._columns.sort()
-
-            if self._columns:
-                new_data = []
-                for column in self._columns:
-                    if column in self._metadata_columns:
-                        metadata_attr = self._metadata_columns[column]
-                        new_value = getattr(event, metadata_attr)
-                    elif column in self._rename_columns:
-                        new_value = data[self._rename_columns[column]]
-                    else:
-                        new_value = data[column]
-                    new_data.append(new_value)
-                data = new_data
-
-        elif isinstance(data, list) and self._columns and self._metadata_columns:
-            new_data = []
-            data_cursor = 0
-            for column in self._columns:
-                if column in self._metadata_columns:
-                    metadata_attr = self._metadata_columns[column]
-                    new_value = getattr(event, metadata_attr)
-                    new_data.append(new_value)
-                else:
-                    new_data.append(data[data_cursor])
-                    data_cursor += 1
-            data = new_data
+                self._infer_columns_from_data = False
+            data = []
+            self._get_column_data_from_dict(data, event, self._columns, self._metadata_columns, self._rename_columns)
```

Comment:
In case you are inferring you can just get `list(data.values())` rather than do all the `_get_column_data_from_dict` logic